# Segmenting and Clustering Neighborhoods in Toronto part I
1.Create Jupiter Notebook & import libraries

In [156]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans
import random
import csv
# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

In [324]:
print('Libraries imported.')

Libraries imported.


In [4]:
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: done

# All requested packages already installed.



In [325]:
import folium

In [326]:
from bs4 import BeautifulSoup

2.Scrape Wikipedia page

In [327]:
data = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
soup = BeautifulSoup(data, 'html.parser')


In [328]:
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"88c1650a-329d-4919-8522-bbc7e0e91068","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":960187814,"wgRevisionId":960187814,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toron

In [329]:
right_table=soup.find('table',class_='wikitable sortable')
right_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td>

In [330]:
all_tables = soup.find_all('table')
all_tables


[<table class="wikitable sortable">
 <tbody><tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighborhood
 </th></tr>
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>
 <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>
 <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>
 <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>
 <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue, Humber Valley Village
 </td></tr>
 <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>
 <tr>
 <td>M2B


3.Create dataframe

* The dataframe consist of three columns: Postcode, Borough, Neighborhood

In [331]:
postcode = []
borough = []
neighborhood = []

In [332]:
for row in right_table.findAll('tr'):
    
    cells= row.findAll('td')
    
    if len(cells) ==3:
        postcode.append(cells[0].find(text=True))
        borough.append(cells[1].find(text=True))
        neighborhood.append(cells[2].find(text=True))
        
        

In [333]:
df = pd.DataFrame({'Postcode': postcode,
                       'Borough': borough,
                       'Neighborhood': neighborhood})

In [334]:
df.head()

Postcode            Borough                Neighborhood
0     M1A
      Not assigned
               Not assigned

1     M2A
      Not assigned
               Not assigned

2     M3A
        North York
                  Parkwoods

3     M4A
        North York
           Victoria Village

4     M5A
  Downtown Toronto
  Regent Park, Harbourfront

* Drop Non assigned values in 'Borough' column

In [335]:
df.drop(df[df.Borough.str.contains('Not assigned')].index,inplace =True)

In [336]:
df = df.reset_index(drop=True)

In [337]:
df.head()

Postcode            Borough                                  Neighborhood
0     M3A
        North York
                                    Parkwoods

1     M4A
        North York
                             Victoria Village

2     M5A
  Downtown Toronto
                    Regent Park, Harbourfront

3     M6A
        North York
             Lawrence Manor, Lawrence Heights

4     M7A
  Downtown Toronto
  Queen's Park, Ontario Provincial Government

* Grouping Neighborhood

In [338]:
df.groupby(['Postcode','Borough'], as_index=False).agg(lambda x: ",".join (x))

Postcode            Borough  \
0       M1B
       Scarborough
   
1       M1C
       Scarborough
   
2       M1E
       Scarborough
   
3       M1G
       Scarborough
   
4       M1H
       Scarborough
   
5       M1J
       Scarborough
   
6       M1K
       Scarborough
   
7       M1L
       Scarborough
   
8       M1M
       Scarborough
   
9       M1N
       Scarborough
   
10      M1P
       Scarborough
   
11      M1R
       Scarborough
   
12      M1S
       Scarborough
   
13      M1T
       Scarborough
   
14      M1V
       Scarborough
   
15      M1W
       Scarborough
   
16      M1X
       Scarborough
   
17      M2H
        North York
   
18      M2J
        North York
   
19      M2K
        North York
   
20      M2L
        North York
   
21      M2M
        North York
   
22      M2N
        North York
   
23      M2P
        North York
   
24      M2R
        North York
   
25      M3A
        North York
   
26      M3B
        North York
   
27      M3C
        North York
   
28      M3H
        North York
   
29      M3J
        North York
   
30      M3K
        North York
   
31      M3L
        North York
   
32      M3M
        North York
   
33      M3N
        North York
   
34      M4A
        North York
   
35      M4B
         East York
   
36      M4C
         East York
   
37      M4E
      East Toronto
   
38      M4G
         East York
   
39      M4H
         East York
   
40      M4J
         East York
   
41      M4K
      East Toronto
   
42      M4L
      East Toronto
   
43      M4M
      East Toronto
   
44      M4N
   Central Toronto
   
45      M4P
   Central Toronto
   
46      M4R
   Central Toronto
   
47      M4S
   Central Toronto
   
48      M4T
   Central Toronto
   
49      M4V
   Central Toronto
   
50      M4W
  Downtown Toronto
   
51      M4X
  Downtown Toronto
   
52      M4Y
  Downtown Toronto
   
53      M5A
  Downtown Toronto
   
54      M5B
  Downtown Toronto
   
55      M5C
  Downtown Toronto
   
56      M5E
  Downtown Toronto
   
57      M5G
  Downtown Toronto
   
58      M5H
  Downtown Toronto
   
59      M5J
  Downtown Toronto
   
60      M5K
  Downtown Toronto
   
61      M5L
  Downtown Toronto
   
62      M5M
        North York
   
63      M5N
   Central Toronto
   
64      M5P
   Central Toronto
   
65      M5R
   Central Toronto
   
66      M5S
  Downtown Toronto
   
67      M5T
  Downtown Toronto
   
68      M5V
  Downtown Toronto
   
69      M5W
  Downtown Toronto
   
70      M5X
  Downtown Toronto
   
71      M6A
        North York
   
72      M6B
        North York
   
73      M6C
              York
   
74      M6E
              York
   
75      M6G
  Downtown Toronto
   
76      M6H
      West Toronto
   
77      M6J
      West Toronto
   
78      M6K
      West Toronto
   
79      M6L
        North York
   
80      M6M
              York
   
81      M6N
              York
   
82      M6P
      West Toronto
   
83      M6R
      West Toronto
   
84      M6S
      West Toronto
   
85      M7A
  Downtown Toronto
   
86      M7R
       Mississauga
   
87      M7Y
      East Toronto
   
88      M8V
         Etobicoke
   
89      M8W
         Etobicoke
   
90      M8X
         Etobicoke
   
91      M8Y
         Etobicoke
   
92      M8Z
         Etobicoke
   
93      M9A
         Etobicoke
   
94      M9B
         Etobicoke
   
95      M9C
         Etobicoke
   
96      M9L
        North York
   
97      M9M
        North York
   
98      M9N
              York
   
99      M9P
         Etobicoke
   
100     M9R
         Etobicoke
   
101     M9V
         Etobicoke
   
102     M9W
         Etobicoke
   

                                          Neighborhood  
0                                     Malvern, Rouge\n  
1             Rouge Hill, Port Union, Highland Creek\n  
2                  Guildwood, Morningside, West Hill\n  
3                                             Woburn\n  
4                                          Cedarbrae\n  
5                    

* Checking Non assigned value for Neighborhood

In [339]:
df.Neighborhood.str.contains('Not assigned')

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
30     False
31     False
32     False
33     False
34     False
35     False
36     False
37     False
38     False
39     False
40     False
41     False
42     False
43     False
44     False
45     False
46     False
47     False
48     False
49     False
50     False
51     False
52     False
53     False
54     False
55     False
56     False
57     False
58     False
59     False
60     False
61     False
62     False
63     False
64     False
65     False
66     False
67     False
68     False
69     False
70     False
71     False
72     False
73     False
74     False
75     False
76     False

* use the .shape method to print the number of rows of your dataframe

In [121]:
df.shape

(103, 3)

# Segmenting and Clustering Neighborhoods in Toronto part II

* geting the latitude and the longitude coordinates of each neighborhood

In [340]:
url = 'http://cocl.us/Geospatial_data'

In [341]:
gecod = pd.read_csv(url)

In [316]:
gecod.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

* useing the Geocoder package or the csv file to create the following dataframe

In [345]:
df['Latitude'] = gecod['Latitude']
df['Longitude'] = gecod['Longitude']

In [346]:
df.head()

Postcode            Borough                                  Neighborhood  \
0     M3A
        North York
                                    Parkwoods
   
1     M4A
        North York
                             Victoria Village
   
2     M5A
  Downtown Toronto
                    Regent Park, Harbourfront
   
3     M6A
        North York
             Lawrence Manor, Lawrence Heights
   
4     M7A
  Downtown Toronto
  Queen's Park, Ontario Provincial Government
   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

# Segmenting and Clustering Neighborhoods in Toronto part III

Explore and cluster the neighborhoods in Toronto

Useing  geopy library to get the latitude and longitude values of Toronto.

In [351]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Create a map of Toronto with neighborhoods superimposed on top.

In [357]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

utilizing the Foursquare API to explore the neighborhoods and segment them

In [358]:
CLIENT_ID = 'FOG4IUBUX0UL2ZPE0YSHWNS41D03YSNNUTINAEZ5XNF2TWUO' # your Foursquare ID
CLIENT_SECRET = 'VHXKEZ4PQALIUCE4RCP13DM5SELIXT1Z0IWAVINUUWLXAIIE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: FOG4IUBUX0UL2ZPE0YSHWNS41D03YSNNUTINAEZ5XNF2TWUO
CLIENT_SECRET:VHXKEZ4PQALIUCE4RCP13DM5SELIXT1Z0IWAVINUUWLXAIIE


Geting the neighborhood's latitude and longitude values

In [370]:
def getNearbyVenues(names, latitudes, longitudes, radius=200, LIMIT=300):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

create a new dataframe called toronto_venues

In [371]:
# type your answer here

toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )


Parkwoods

Victoria Village

Regent Park, Harbourfront

Lawrence Manor, Lawrence Heights

Queen's Park, Ontario Provincial Government

Islington Avenue, Humber Valley Village

Malvern, Rouge

Don Mills

Parkview Hill, Woodbine Gardens

Garden District, Ryerson

Glencairn

West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale

Rouge Hill, Port Union, Highland Creek

Don Mills

Woodbine Heights

St. James Town

Humewood-Cedarvale

Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood

Guildwood, Morningside, West Hill

The Beaches

Berczy Park

Caledonia-Fairbanks

Woburn

Leaside

Central Bay Street

Christie

Cedarbrae

Hillcrest Village

Bathurst Manor, Wilson Heights, Downsview North

Thorncliffe Park

Richmond, Adelaide, King

Dufferin, Dovercourt Village

Scarborough Village

Fairview, Henry Farm, Oriole

Northwood Park, York University

East Toronto, Broadview North (Old East York)

Harbourfront East, Union Station, Toronto Islands

Little Portugal, Trinit

checking the size of the resulting dataframe

In [372]:
print(toronto_venues.shape)
toronto_venues.head()

(526, 7)


Neighborhood  Neighborhood Latitude  \
0         Lawrence Manor, Lawrence Heights
              43.770992   
1  Islington Avenue, Humber Valley Village
              43.744734   
2                 Garden District, Ryerson
              43.692657   
3                                Glencairn
              43.757410   
4                           St. James Town
              43.799525   

   Neighborhood Longitude                          Venue  Venue Latitude  \
0              -79.216917             Korean Grill House       43.770812   
1              -79.239476                   McCowan Park       43.745089   
2              -79.264848                 The Birchcliff       43.691666   
3              -79.273304                Scarborough LRT       43.756465   
4              -79.318389  Mr Congee Chinese Cuisine 龍粥記       43.798879   

   Venue Longitude      Venue Category  
0       -79.214502   Korean Restaurant  
1       -79.239336          Playground  
2       -79.264532                Café  
3       -79.272194  Light Rail Station  
4       -79.318335  Chinese Restaurant

 checking how many venues were returned for each neighborhood

In [373]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               1   
                                                                        7   
Bathurst Manor, Wilson Heights, Downsview North                        14   
                                                                        8   
Bayview Village                                                        27   
                                                                       16   
Bedford Park, Lawrence Manor East                                       2   
                                                                        4   
Birch Cliff, Cliffside West                                             2   
                                                                       13   
Brockton, Parkdale Village, Exhibition Place                            1   
                                                                        2   
CN Tower, King and Spadina, Railway Lands, Harb...                     31   
Canada Post Gateway Processing Centre                                  55   
                                                                        2   
Church and Wellesley                                                    4   
                                                                        1   
Cliffside, Cliffcrest, Scarborough Village West                         1   
                                                                        1   
Commerce Court, Victoria Hotel                                          1   
                                                                        1   
Davisville                                                             21   
                                                                       22   
Davisville North                                                       14   
                                                                        1   
Downsview                                                              13   
                                                                       16   
Dufferin, Dovercourt Village                                            1   
                                                                        6   
East Toronto, Broadview North (Old East York)                           1   
                                                                       18   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      1   
Forest Hill North & West, Forest Hill Road Park                        45   
                                                                        3   
Garden District, Ryerson                                                2   
                                                                        1   
Glencairn                                                               2   
                                                                        4   
Golden Mile, Clairlea, Oakridge                                         2   
                                                                        2   
Guildwood, Morningside, West Hill                                       1   
                                                                        7   
High Park, The Junction South                                           1   
                                                                        3   
Hillcrest Village                                                      13   
                                                                        1   
Humber Summit                                                          31   
                                                                        1   
Humberlea, Emery                                                        2   
                                                                        3   
India Bazaar, The Beaches West                  

In [374]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 154 uniques categories.


Analyze Each Neighborhood

In [375]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  American Restaurant  \
0            0                  0               0                    0   
1            0                  0               0                    0   
2            0                  0               0                    0   
3            0                  0               0                    0   
4            0                  0               0                    0   

   Arepa Restaurant  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0                 0            0                    0                 0   
1                 0            0                    0                 0   
2                 0            0                    0                 0   
3                 0            0                    0                 0   
4                 0            0                    0                 0   

   Astrologer  Auto Workshop  BBQ Joint  Bakery  Bank  Bar  Beer Bar  \
0           0              0          0       0     0    0         0   
1           0              0          0       0     0    0         0   
2           0              0          0       0     0    0         0   
3           0              0          0       0     0    0         0   
4           0              0          0       0     0    0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bookstore  Boutique  \
0           0                   0          0          0         0   
1           0                   0          0          0         0   
2           0                   0          0          0         0   
3           0                   0          0          0         0   
4           0                   0          0          0         0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Stop  Café  Camera Store  Caribbean Restaurant  \
0              0         0     0             0                     0   
1              0         0     0             0                     0   
2              0         0     1             0                     0   
3              0         0     0             0                     0   
4              0         0     0             0                     0   

   Cheese Shop  Chinese Restaurant  Chocolate Shop  Clothing Store  \
0            0                   0               0               0   
1            0                   0               0               0   
2            0                   0               0               0   
3            0                   0               0               0   
4            0                   1               0               0   

   Cocktail Bar  Coffee Shop  College Gym  College Rec Center  \
0             0            0            0                   0   
1             0            0            0                   0   
2             0            0            0                   0   
3             0            0            0                   0   
4             0            0            0                   0   

   Colombian Restaurant  Comfort Food Restaurant  Comic Shop  Concert Hall  \
0                     0                        0           0             0   
1                     0                        0           0             0   
2                     0                        0           0             0   
3                     0                        0           0             0   
4                     0                        0           0             0   

   Convenience Store  Costume Shop  Creperie  Cuban Restaurant  Cupcake Shop  \
0                  0             0    

group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [378]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                          Agincourt
     0.000000   
1    Bathurst Manor, Wilson Heights, Downsview North
     0.000000   
2                                    Bayview Village
     0.071429   
3                  Bedford Park, Lawrence Manor East
     0.000000   
4                        Birch Cliff, Cliffside West
     0.000000   
5       Brockton, Parkdale Village, Exhibition Place
     0.000000   
6   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
7              Canada Post Gateway Processing Centre
     0.000000   
8                               Church and Wellesley
     0.000000   
9    Cliffside, Cliffcrest, Scarborough Village West
     0.000000   
10                    Commerce Court, Victoria Hotel
     0.000000   
11                                        Davisville
     0.000000   
12                                  Davisville North
     0.000000   
13                                         Downsview
     0.000000   
14                      Dufferin, Dovercourt Village
     0.000000   
15     East Toronto, Broadview North (Old East York)
     0.000000   
16  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
17   Forest Hill North & West, Forest Hill Road Park
     0.000000   
18                          Garden District, Ryerson
     0.000000   
19                                         Glencairn
     0.000000   
20                   Golden Mile, Clairlea, Oakridge
     0.000000   
21                 Guildwood, Morningside, West Hill
     0.000000   
22                     High Park, The Junction South
     0.000000   
23                                 Hillcrest Village
     0.000000   
24                                     Humber Summit
     0.000000   
25                                  Humberlea, Emery
     0.000000   
26                    India Bazaar, The Beaches West
     0.000000   
27           Islington Avenue, Humber Valley Village
     0.000000   
28       Kennedy Park, Ionview, East Birchmount Park
     0.000000   
29         Kensington Market, Chinatown, Grange Park
     0.000000   
30  Kingsview Village, St. Phillips, Martin Grove ...     0.055556   
31                  Lawrence Manor, Lawrence Heights
     0.000000   
32                                     Lawrence Park
     0.000000   
33                                           Leaside
     0.000000   
34                          Little Portugal, Trinity
     0.000000   
35  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
36                       Moore Park, Summerhill East
     0.000000   
37          North Park, Maple Leaf Park, Upwood Park
     0.000000   
38                North Toronto West,  Lawrence Park
     0.000000   
39                   Northwood Park, York University
     0.000000   
40  Old Mill South, King's Mill Park, Sunnylea, Hu...     0.000000   
41                                          Roselawn
     0.000000   
42                     Runnymede, The Junction North
     0.000000   
43  South Steeles, Silverstone, Humbergate, Jamest...     0.000000   
44                                    St. James Town
     0.000000   
45                       St. James Town, Cabbagetown
     0.000000   
46                                   Studio District
     0.000000   
47  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
48          Toronto Dominion Centre, Design Exchange
     0.000000   
49                    University of Toronto, Harbord
     0.000000   
50                                       Upper Rouge
     0.000000   
51                                         Westmount
     0.000000   
52                                            Weston
     0.000000   
53                                 Wexford, Maryvale
     0.000000   
54                           Willowdale, Newtonbrook
     0.000000   
55                       Willowdale, Willowdale East
     0.000000   
56                       Willowdale, Willowdale Wes

confirming new size

In [379]:
toronto_grouped.shape

(59, 154)

printing each neighborhood along with the top 5 most common venues

In [380]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt
----
                      venue  freq
0                      Café   1.0
1               Yoga Studio   0.0
2                    Museum   0.0
3  Mediterranean Restaurant   0.0
4        Mexican Restaurant   0.0


----Bathurst Manor, Wilson Heights, Downsview North
----
                       venue  freq
0  Middle Eastern Restaurant  0.14
1                Pizza Place  0.14
2                Coffee Shop  0.14
3              Deli / Bodega  0.14
4        Fried Chicken Joint  0.14


----Bayview Village
----
                 venue  freq
0    Indian Restaurant  0.14
1       Sandwich Place  0.14
2          Yoga Studio  0.07
3          Supermarket  0.07
4  Housing Development  0.07


----Bedford Park, Lawrence Manor East
----
                 venue  freq
0  Japanese Restaurant  0.25
1            Gastropub  0.12
2             Creperie  0.12
3          Coffee Shop  0.12
4           Poke Place  0.12


----Birch Cliff, Cliffside West
----
              venue  freq
0       Coffee Shop  0.

put that into a pandas dataframe

In [381]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

 creating the new dataframe and display the top 10 venues for each neighborhood.

In [383]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt
                  Café   
1  Bathurst Manor, Wilson Heights, Downsview North
           Coffee Shop   
2                                  Bayview Village
     Indian Restaurant   
3                Bedford Park, Lawrence Manor East
   Japanese Restaurant   
4                      Birch Cliff, Cliffside West
            Steakhouse   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0              Women's Store     Electronics Store   Dumpling Restaurant   
1  Middle Eastern Restaurant           Pizza Place            Restaurant   
2             Sandwich Place           Yoga Studio           Supermarket   
3                Coffee Shop             Gastropub              Creperie   
4                Coffee Shop      Asian Restaurant                Garden   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Discount Store                 Diner    Dim Sum Restaurant   
1         Deli / Bodega   Fried Chicken Joint      Sushi Restaurant   
2           Gas Station   Housing Development          Intersection   
3            Poke Place    Italian Restaurant          Camera Store   
4          Concert Hall          Noodle House           Opera House   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dessert Shop      Department Store          Deli / Bodega  
1                      Diner    Dim Sum Restaurant           Dessert Shop  
2  Middle Eastern Restaurant              Pharmacy            Pizza Place  
3              Women's Store      Department Store         Discount Store  
4                       Park              Pharmacy    Japanese Restaurant

Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [384]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 1, 1, 1, 1, 1, 1, 1, 4, 1], dtype=int32)

 createing a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [389]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postcode            Borough                                  Neighborhood  \
0     M3A
        North York
                                    Parkwoods
   
1     M4A
        North York
                             Victoria Village
   
2     M5A
  Downtown Toronto
                    Regent Park, Harbourfront
   
3     M6A
        North York
             Lawrence Manor, Lawrence Heights
   
4     M7A
  Downtown Toronto
  Queen's Park, Ontario Provincial Government
   

    Latitude  Longitude  Cluster_Labels  Cluster Labels 1st Most Common Venue  \
0  43.806686 -79.194353             NaN             NaN                   NaN   
1  43.784535 -79.160497             NaN             NaN                   NaN   
2  43.763573 -79.188711             NaN             NaN                   NaN   
3  43.770992 -79.216917             1.0             1.0     Korean Restaurant   
4  43.773136 -79.239476             NaN             NaN                   NaN   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1                   NaN                   NaN                   NaN   
2                   NaN                   NaN                   NaN   
3      Department Store  Ethiopian Restaurant     Electronics Store   
4                   NaN                   NaN                   NaN   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   NaN                   NaN                   NaN   
1                   NaN                   NaN                   NaN   
2                   NaN                   NaN                   NaN   
3   Dumpling Restaurant        Discount Store                 Diner   
4                   NaN                   NaN                   NaN   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   NaN                   NaN                    NaN  
1                   NaN                   NaN                    NaN  
2                   NaN                   NaN                    NaN  
3    Dim Sum Restaurant          Dessert Shop          Women's Store  
4                   NaN                   NaN                    NaN

 visualizing the resulting clusters

In [394]:
toronto_merged.dropna(inplace =True)


In [395]:
toronto_merged.head()

Postcode            Borough                              Neighborhood  \
3      M6A
        North York
         Lawrence Manor, Lawrence Heights
   
5      M9A
         Etobicoke
  Islington Avenue, Humber Valley Village
   
9      M5B
  Downtown Toronto
                 Garden District, Ryerson
   
10     M6B
        North York
                                Glencairn
   
15     M5C
  Downtown Toronto
                           St. James Town
   

     Latitude  Longitude  Cluster_Labels  Cluster Labels  \
3   43.770992 -79.216917             1.0             1.0   
5   43.744734 -79.239476             1.0             1.0   
9   43.692657 -79.264848             2.0             2.0   
10  43.757410 -79.273304             1.0             1.0   
15  43.799525 -79.318389             1.0             1.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
3      Korean Restaurant      Department Store  Ethiopian Restaurant   
5             Playground         Women's Store          Cupcake Shop   
9                   Café         Women's Store     Electronics Store   
10    Light Rail Station      Department Store  Ethiopian Restaurant   
15    Chinese Restaurant  Fast Food Restaurant         Grocery Store   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
3      Electronics Store   Dumpling Restaurant        Discount Store   
5    Dumpling Restaurant        Discount Store                 Diner   
9    Dumpling Restaurant        Discount Store                 Diner   
10     Electronics Store   Dumpling Restaurant        Discount Store   
15                  Bank           Coffee Shop              Pharmacy   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
3                  Diner    Dim Sum Restaurant          Dessert Shop   
5     Dim Sum Restaurant          Dessert Shop      Department Store   
9     Dim Sum Restaurant          Dessert Shop      Department Store   
10                 Diner    Dim Sum Restaurant          Dessert Shop   
15           Pizza Place        Sandwich Place           Supermarket   

   10th Most Common Venue  
3           Women's Store  
5           Deli / Bodega  
9           Deli / Bodega  
10          Women's Store  
15           Noodle House

In [398]:
toronto_merged.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 62 entries, 3 to 101
Data columns (total 17 columns):
Postcode                  62 non-null object
Borough                   62 non-null object
Neighborhood              62 non-null object
Latitude                  62 non-null float64
Longitude                 62 non-null float64
Cluster_Labels            62 non-null float64
Cluster Labels            62 non-null float64
1st Most Common Venue     62 non-null object
2nd Most Common Venue     62 non-null object
3rd Most Common Venue     62 non-null object
4th Most Common Venue     62 non-null object
5th Most Common Venue     62 non-null object
6th Most Common Venue     62 non-null object
7th Most Common Venue     62 non-null object
8th Most Common Venue     62 non-null object
9th Most Common Venue     62 non-null object
10th Most Common Venue    62 non-null object
dtypes: float64(4), object(13)
memory usage: 8.7+ KB


In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

 Examine Clusters

Cluster 1

In [401]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster_Labels  Cluster Labels  1st Most Common Venue  \
68  Central Toronto
             0.0             0.0  Performing Arts Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
68         Deli / Bodega     Electronics Store   Dumpling Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
68        Discount Store                 Diner    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
68          Dessert Shop      Department Store           Cupcake Shop

Cluster 2

In [405]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels  Cluster Labels  \
3          North York
             1.0             1.0   
5           Etobicoke
             1.0             1.0   
10         North York
             1.0             1.0   
15   Downtown Toronto
             1.0             1.0   
17          Etobicoke
             1.0             1.0   
18        Scarborough
             1.0             1.0   
27         North York
             1.0             1.0   
28         North York
             1.0             1.0   
31       West Toronto
             1.0             1.0   
34         North York
             1.0             1.0   
35          East York
             1.0             1.0   
38        Scarborough
             1.0             1.0   
39         North York
             1.0             1.0   
40         North York
             1.0             1.0   
43       West Toronto
             1.0             1.0   
44        Scarborough
             1.0             1.0   
45         North York
             1.0             1.0   
46         North York
             1.0             1.0   
47       East Toronto
             1.0             1.0   
49         North York
             1.0             1.0   
50         North York
             1.0             1.0   
51        Scarborough
             1.0             1.0   
52         North York
             1.0             1.0   
53         North York
             1.0             1.0   
54       East Toronto
             1.0             1.0   
55         North York
             1.0             1.0   
57         North York
             1.0             1.0   
58        Scarborough
             1.0             1.0   
59         North York
             1.0             1.0   
60         North York
             1.0             1.0   
61    Central Toronto
             1.0             1.0   
62    Central Toronto
             1.0             1.0   
63               York
             1.0             1.0   
64               York
             1.0             1.0   
66         North York
             1.0             1.0   
67    Central Toronto
             1.0             1.0   
69       West Toronto
             1.0             1.0   
70          Etobicoke
             1.0             1.0   
71        Scarborough
             1.0             1.0   
72         North York
             1.0             1.0   
73    Central Toronto
             1.0             1.0   
76        Mississauga
             1.0             1.0   
77          Etobicoke
             1.0             1.0   
79    Central Toronto
             1.0             1.0   
80   Downtown Toronto
             1.0             1.0   
83    Central Toronto
             1.0             1.0   
84   Downtown Toronto
             1.0             1.0   
85        Scarborough
             1.0             1.0   
87   Downtown Toronto
             1.0             1.0   
95        Scarborough
             1.0             1.0   
101         Etobicoke
             1.0             1.0   

       1st Most Common Venue      2nd Most Common Venue 3rd Most Common Venue  \
3          Korean Restaurant           Department Store  Ethiopian Restaurant   
5                 Playground              Women's Store          Cupcake Shop   
10        Light Rail Station           Department Store  Ethiopian Restaurant   
15        Chinese Restaurant       Fast Food Restaurant         Grocery Store   
17               Golf Course              Women's Store          Cupcake Shop   
18            Clothing Store              Women's Store           Coffee Shop   
27                Restaurant           Asian Restaurant         Shopping Mall   
28               Coffee Shop  Middle Eastern Restaurant           Pizza Place   
31               Wings Joint               Home Service         Deli / Bodega   
34         French Restaurant               Intersection         Women's Store   
35            Breakfast Spot                       Café             Pet Store   
38            Sandwich Pla

Cluster 3

In [414]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels  Cluster Labels 1st Most Common Venue  \
9   Downtown Toronto
             2.0             2.0                  Café   
78       Scarborough
             2.0             2.0                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9          Women's Store     Electronics Store   Dumpling Restaurant   
78         Women's Store     Electronics Store   Dumpling Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9         Discount Store                 Diner    Dim Sum Restaurant   
78        Discount Store                 Diner    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9           Dessert Shop      Department Store          Deli / Bodega  
78          Dessert Shop      Department Store          Deli / Bodega

Cluster 4

In [415]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels  Cluster Labels 1st Most Common Venue  \
23         East York
             3.0             3.0                  Park   
37      West Toronto
             3.0             3.0                  Park   
42  Downtown Toronto
             3.0             3.0                  Park   
48  Downtown Toronto
             3.0             3.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23     Convenience Store         Women's Store      Department Store   
37                 Trail          Cupcake Shop   Dumpling Restaurant   
42     Fish & Chips Shop  Ethiopian Restaurant     Electronics Store   
48         Women's Store         Deli / Bodega     Electronics Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23     Electronics Store   Dumpling Restaurant        Discount Store   
37        Discount Store                 Diner    Dim Sum Restaurant   
42   Dumpling Restaurant        Discount Store                 Diner   
48   Dumpling Restaurant        Discount Store                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23                 Diner    Dim Sum Restaurant           Dessert Shop  
37          Dessert Shop      Department Store          Deli / Bodega  
42    Dim Sum Restaurant          Dessert Shop       Department Store  
48    Dim Sum Restaurant          Dessert Shop       Department Store

Cluster 5

In [416]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels  Cluster Labels 1st Most Common Venue  \
86   Central Toronto
             4.0             4.0           Coffee Shop   
89         Etobicoke
             4.0             4.0           Coffee Shop   
96  Downtown Toronto
             4.0             4.0           Pizza Place   
99  Downtown Toronto
             4.0             4.0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
86         Women's Store         Deli / Bodega     Electronics Store   
89              Pharmacy           Pizza Place          Cupcake Shop   
96         Women's Store          Cupcake Shop   Dumpling Restaurant   
99           Pizza Place         Women's Store         Deli / Bodega   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
86   Dumpling Restaurant        Discount Store                 Diner   
89        Discount Store                 Diner    Dim Sum Restaurant   
96        Discount Store                 Diner    Dim Sum Restaurant   
99   Dumpling Restaurant        Discount Store                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
86    Dim Sum Restaurant          Dessert Shop       Department Store  
89          Dessert Shop      Department Store          Deli / Bodega  
96          Dessert Shop      Department Store          Deli / Bodega  
99    Dim Sum Restaurant          Dessert Shop       Department Store

Obserwation
On the base above analysis we can see that 526 venues return for each nighborhoods. The larges areas are: Downsview 55 venues 10% of tolal, Lowrence Park 45 venues 8.5% of tolal , Westmount 35 venues 6.6% of total.
Data contains 154 unik categories, looking at frequensy of visitors Leasge venue Park has the highest resoult 0.67, Humberlea, Emery venue Coffe Shop 0.38, University of Toronto, Harbord venue Museum 0.33

Cluster examination
Cluster2 is the bigest w hile Cluster one has only one neighborhood 